From 6ec19fab6d344d9679305e3df511b6d51d5fd353 Mon Sep 17 00:00:00 2001
From: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
Date: Thu, 14 Nov 2024 12:28:05 +0530
Subject: [PATCH] UPSTREAM:DP:wifi:ath12k: Prefetch hw and sw desc in tx
 completion handler

Prefetch n+3 and consective hw and sw desc in tx completion
handler to improve CPU optimization

This change gives 50-70mbps KPI improvement in DL traffic

Patch Dependency: None

Signed-off-by: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 62 +++++++++++++++++++++++--
 drivers/net/wireless/ath/ath12k/hal.c   |  8 +---
 2 files changed, 58 insertions(+), 12 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index 10527ca..eafa9d4 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -1574,6 +1574,33 @@ static inline void ath12k_dp_tx_decrement_by_count(struct ath12k *ar,
 	}
 }
 
+struct hal_wbm_completion_ring_tx *
+ath12k_hal_srng_dst_prefetch_tx_status(struct hal_srng *srng)
+{
+	u32 *desc, tp, entry_size;
+
+	lockdep_assert_held(&srng->lock);
+	tp = srng->u.dst_ring.tp;
+	entry_size = srng->entry_size;
+
+	if (tp == srng->u.dst_ring.cached_hp)
+		return NULL;
+	desc = srng->ring_base_vaddr + tp;
+
+	/* check if the descriptor address is 64byte aligned address */
+	if (!IS_ALIGNED((unsigned long)desc, 64))
+		tp += entry_size;
+	else
+		tp += (entry_size * 2);
+
+	if (tp == srng->ring_size)
+		tp = 0;
+
+	desc = srng->ring_base_vaddr + tp;
+	prefetch(desc);
+	return (struct hal_wbm_completion_ring_tx *)(desc + entry_size);
+}
+
 int ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id,
 				    int budget)
 {
@@ -1583,7 +1610,7 @@ int ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id,
 	int hal_ring_id = tx_ring->tcl_comp_ring.ring_id;
 	int data_ring_id = tx_ring->tcl_data_ring_id;
 	struct hal_srng *status_ring = &ab->hal.srng_list[hal_ring_id];
-	struct ath12k_tx_desc_info *tx_desc = NULL;
+	struct ath12k_tx_desc_info *tx_desc, *last_pf_sw_desc = NULL;
 	struct list_head desc_free_list, *cur;
 	struct sk_buff *msdu;
 	u8 mac_id;
@@ -1592,10 +1619,10 @@ int ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id,
 	struct ath12k_skb_cb *skb_cb;
 #endif
 	u64 desc_va;
-	u32 desc_id;
+	u32 desc_id, tp;
 	int valid_entries;
 	int orig_budget = budget;
-	struct hal_wbm_completion_ring_tx *tx_status;
+	struct hal_wbm_completion_ring_tx *tx_status, *last_pf_hw_desc;
 	enum hal_wbm_rel_src_module buf_rel_source;
 	struct sk_buff_head free_list_head;
 	int dp_tx_dec_count[MAX_RADIOS] = {0};
@@ -1617,6 +1644,7 @@ int ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id,
 
 	ath12k_hal_srng_dst_invalidate_entry(ab, status_ring, valid_entries);
 	skb_queue_head_init(&free_list_head);
+	last_pf_hw_desc = ath12k_hal_srng_dst_prefetch_tx_status(status_ring);
 
 	while (budget && (tx_status = (struct hal_wbm_completion_ring_tx *)ath12k_hal_srng_dst_get_next_cache_entry(ab, status_ring))) {
 		budget--;
@@ -1669,8 +1697,32 @@ int ath12k_dp_tx_completion_handler(struct ath12k_base *ab, int ring_id,
 			continue;
 		}
 
-		prefetch(msdu);
-		prefetch( (u8 *) msdu + 64);
+
+		if (last_pf_sw_desc) {
+			prefetch((u8 *)(last_pf_sw_desc->skb) + offsetof(struct sk_buff, headers));
+		}
+
+		if (valid_entries && last_pf_hw_desc) {
+			desc_va = ((u64)last_pf_hw_desc->buf_va_hi << 32 |
+					last_pf_hw_desc->buf_va_lo);
+			prefetch(desc_va);
+			last_pf_sw_desc = (struct ath12k_tx_desc_info *)((unsigned long)desc_va);
+
+			if (status_ring->u.dst_ring.tp == status_ring->u.dst_ring.cached_hp) {
+				last_pf_hw_desc = NULL;
+			} else {
+				tp = status_ring->u.dst_ring.tp + (status_ring->entry_size * 2);
+
+				if (tp == status_ring->ring_size)
+					tp = 0;
+
+				if (tp != status_ring->u.dst_ring.cached_hp)
+					last_pf_hw_desc = status_ring->ring_base_vaddr + tp;
+
+				if (!IS_ALIGNED((unsigned long)last_pf_hw_desc, 64))
+					prefetch(last_pf_hw_desc);
+			}
+		}
 
 		if (unlikely(buf_rel_source == HAL_WBM_REL_SRC_MODULE_FW)) {
 			ath12k_dp_tx_process_htt_tx_complete(ab,
diff --git a/drivers/net/wireless/ath/ath12k/hal.c b/drivers/net/wireless/ath/ath12k/hal.c
index 378c80c..8cdb06f 100644
--- a/drivers/net/wireless/ath/ath12k/hal.c
+++ b/drivers/net/wireless/ath/ath12k/hal.c
@@ -1873,7 +1873,7 @@ void *ath12k_hal_srng_dst_get_next_entry(struct ath12k_base *ab,
 u32 *ath12k_hal_srng_dst_get_next_cache_entry(struct ath12k_base *ab,
                                               struct hal_srng *srng)
 {
-	u32 *desc, *desc_next;
+	u32 *desc;
 	lockdep_assert_held(&srng->lock);
 
 	if (srng->u.dst_ring.tp == srng->u.dst_ring.cached_hp)
@@ -1887,12 +1887,6 @@ u32 *ath12k_hal_srng_dst_get_next_cache_entry(struct ath12k_base *ab,
 	if (srng->u.dst_ring.tp == srng->ring_size)
 		srng->u.dst_ring.tp = 0;
 
-	/* Try to prefetch the next descriptor in the ring */
-	if (srng->u.dst_ring.tp != srng->u.dst_ring.cached_hp) {
-		/* prefetch only if desc is available */
-		desc_next = srng->ring_base_vaddr + srng->u.dst_ring.tp;
-		prefetch(desc_next);
-	}
 	return desc;
 }
 
-- 
2.34.1

